### PR TITLE
[NFC] Refactor the layering of IREE-specific transform dialect extens…

### PR DESCRIPTION
**Pull request by nicolasvasilache**
…ions and registrations.

This revision allows better isolation and reduced dependencies of transform dialect extensions and registrations.
This is a prerequisite to allow adding LLVMGPU transforms with proper dependence layering.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectExtensions/BUILD [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# Copyright 2019 The IREE Authors
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectExtensions/BUILD:1]: Nit: 2022?

**Review by matthias-springer — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensionsOps.td [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+include "mlir/Dialect/Transform/IR/TransformDialect.td"
+include "mlir/Dialect/Transform/IR/TransformInterfaces.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/IR/OpBase.td"
+
+def FunctionalStyleMultiOperandMultiResultTransformOpTrait
+    : NativeOpTrait<"FunctionalStyleMultiOperandMultiResultTransformOpTrait"> {
+  let cppNamespace = "::mlir::transform";
+}
+
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensionsOps.td:20]: Upstream supports this now.

File: tests/e2e/linalg_transform/benchmark_reduction.mlir [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// func.func @reduce(%arg : tensor<1x128x384xf32>) -> (tensor<1x128xf32>) {
+//   %cst = arith.constant -0.000000e+00 : f32
```
Comment (ftynse) [inline tests/e2e/linalg_transform/benchmark_reduction.mlir:2]: This doesn't look like a `RUN` line.

**Review by ftynse — APPROVED** [review]
I suppose you intended to feel the scaffolding soon, otherwise we get three empty extensions.

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectExtensions/BUILD (hunk 1 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectExtensions/BUILD:1]: yes GPU coming next, CPU is a bit unclear atm re bufferization but it's fine as a placeholder I think

File: tests/e2e/linalg_transform/benchmark_reduction.mlir (hunk 3 above)
Comment (nicolasvasilache) [reply tests/e2e/linalg_transform/benchmark_reduction.mlir:2]: thanks, spurious change with rebasing and all ..

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensionsOps.td (hunk 2 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensionsOps.td:20]: cool, thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensions.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_CODEGEN_LLVMCPU_TRANSFORMDIALECTEXTENSIONS_TRANSFORMDIALECTLLVMCPUEXTENSIONS_H_
+#define IREE_COMPILER_CODEGEN_LLVMCPU_TRANSFORMDIALECTEXTENSIONS_TRANSFORMDIALECTLLVMCPUEXTENSIONS_H_
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensions.h:8]: These file names are too long. CMake on Windows has limitations with long paths and this is our new longest path

```
[cmake] CMake Warning in compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/CMakeLists.txt:
[cmake]   The object file directory
[cmake] 
[cmake]     D:/dev/projects/iree-build/compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/CMakeFiles/iree_compiler_Codegen_LLVMCPU_TransformDialectExtensions_TransformDialectLLVMCPUExtensions.objects.dir/./
[cmake] 
[cmake]   has 213 characters.  The maximum full path to an object file is 250
[cmake]   characters (see CMAKE_OBJECT_PATH_MAX).  Object file
[cmake] 
[cmake]     TransformDialectLLVMCPUExtensions.cpp.obj
[cmake] 
[cmake]   cannot be safely placed under this directory.  The build may not work
[cmake]   correctly.
```

Can you shorten both the directory names (e.g. `TransformExtensions`) and the file names (e.g. remove redundant words, so `Codegen/LLVMCPU/TransformExtensions/LLVMCPUExtensions.h` or similar)?

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensions.h:8]: (@nicolasvasilache )

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TransformDialectExtensions/TransformDialectLLVMCPUExtensions.h:8]: Filed https://github.com/iree-org/iree/issues/9680

**Pull request merged**